File: modules/authentication/pkg/javascript/examples/test_refresh.js
/**
 * Token Refresh Test
 *
 * Tests the token refresh functionality of the authentication client.
 * Similar to examples/refresh.rs but using the JavaScript wrapper.
 */

import { Client } from '../dist/index.js';
import * as dotenv from 'dotenv';

// Load environment variables
dotenv.config({ path: '../../.env' });

async function main() {
  console.log('=== Token Refresh Test ===\n');

  // Configuration
  const config = {
    apiUrl: process.env.API_URL || 'https://api.aukiverse.com',
    refreshUrl: process.env.REFRESH_URL || 'https://api.aukiverse.com/user/refresh',
    ddsUrl: process.env.DDS_URL || 'https://dds.posemesh.org',
    clientId: process.env.CLIENT_ID || 'javascript-test',
    // Set a very high refresh threshold so we can manually trigger refresh
    refreshThresholdMs: 3600000, // 1 hour
  };

  console.log('Configuration:');
  console.log(`  API URL: ${config.apiUrl}`);
  console.log(`  Refresh URL: ${config.refreshUrl}`);
  console.log(`  DDS URL: ${config.ddsUrl}`);
  console.log(`  Client ID: ${config.clientId}`);
  console.log(`  Refresh threshold: ${config.refreshThresholdMs}ms (${config.refreshThresholdMs / 3600000}h)\n`);

  // Credentials
  const credentials = {
    type: 'email',
    email: process.env.EMAIL,
    password: process.env.PASSWORD
  };

  if (!credentials.email || !credentials.password) {
    console.error('❌ EMAIL and PASSWORD must be set in .env file');
    process.exit(1);
  }

  console.log('Using Email/Password authentication');
  console.log(`  Email: ${credentials.email}\n`);

  // Create the authentication client
  const client = await Client.create(config);
  client.setCredentials(credentials);

  try {
    // Step 1: Initial Network Authentication
    console.log('=== Step 1: Initial Network Authentication ===');
    const networkToken = await client.authenticate();
    console.log('✓ Network token obtained!');
    console.log(`  Token: ${networkToken.token.substring(0, 50)}...`);
    console.log(`  Expires at: ${networkToken.expires_at} (${networkToken.expires_at - Date.now()}ms from now)\n`);

    // Step 2: Authenticate to Discovery
    console.log('=== Step 2: Discovery Authentication ===');
    const discoveryToken = await client.authenticateDiscovery();
    console.log('✓ Discovery token obtained!');
    console.log(`  Token: ${discoveryToken.token.substring(0, 50)}...`);
    console.log(`  Expires at: ${discoveryToken.expires_at} (${discoveryToken.expires_at - Date.now()}ms from now)\n`);

    // Step 3: Get Domain Access
    const domainId = process.env.DOMAIN_ID;
    if (!domainId) {
      console.log('⚠ DOMAIN_ID not set in .env, skipping domain access step\n');
    } else {
      console.log('=== Step 3: Domain Access ===');
      console.log(`Requesting access to domain: ${domainId}`);
      const domainAccess = await client.getDomainAccess(domainId);
      console.log('✓ Domain access obtained!');
      console.log(`  Domain: ${domainAccess.name} (${domainAccess.id})`);
      console.log(`  Access token: ${domainAccess.access_token.substring(0, 50)}...`);
      console.log(`  Expires at: ${domainAccess.expires_at} (${domainAccess.expires_at - Date.now()}ms from now)\n`);
    }

    // Step 4: Test Token Refresh
    console.log('=== Step 4: Testing Token Refresh ===');
    console.log('Note: Only the network token has a refresh_token field.');
    console.log('Discovery and domain tokens must be re-requested, not refreshed.\n');

    console.log('Calling authenticate() again to test refresh logic...');

    // Get the current network token to compare
    const networkTokenBefore = client.getNetworkToken();
    console.log(`Network token before: ${networkTokenBefore.token.substring(0, 30)}...`);

    // Try to authenticate again
    // Note: The JavaScript wrapper automatically executes actions and throws "No response"
    // if no actions are returned (token still valid). This is different from the Rust client
    // where you check actions.is_empty() before executing.
    try {
      const networkToken2 = await client.authenticate();
      console.log(`Network token after: ${networkToken2.token.substring(0, 30)}...`);
      console.log('✓ Token was refreshed or re-authenticated!\n');
    } catch (error) {
      // If error message is "No response", it means no actions were returned by the core,
      // which means the token is still valid and doesn't need refresh
      if (error.message && error.message.includes('No response')) {
        console.log('✓ No refresh needed - token is still valid and not near expiry\n');
      } else {
        // Real error
        console.error(`Unexpected error: ${error.message}`);
        throw error;
      }
    }

    // Step 5: Test Discovery Token Re-request
    console.log('=== Step 5: Testing Discovery Token Re-request ===');
    console.log('Calling authenticateDiscovery() again...');

    try {
      const discoveryToken2 = await client.authenticateDiscovery();
      console.log('✓ Discovery token was re-requested!\n');
    } catch (error) {
      // If error message is "No response", token is still valid
      if (error.message && error.message.includes('No response')) {
        console.log('✓ No re-request needed - discovery token is still valid\n');
      } else {
        console.error(`Unexpected error: ${error.message}`);
        throw error;
      }
    }

    // Step 6: Simulate Token Expiration
    console.log('=== Step 6: Simulating Token Expiration ===');
    console.log('Simulating that 1 hour has passed (tokens should be expired)...\n');

    const now = Date.now();
    const futureTime = now + 3660000; // 61 minutes

    console.log(`Current time: ${now}`);
    console.log(`Simulated time: ${futureTime} (${futureTime - now}ms in the future)`);
    console.log(`Is authenticated: ${client.isAuthenticated()}\n`);

    // Check which tokens are expired (relative to future time)
    const networkTokenCurrent = client.getNetworkToken();
    if (networkTokenCurrent) {
      const expired = networkTokenCurrent.expires_at < futureTime;
      console.log(`Network token expired at simulated time: ${expired}`);
      console.log(`  Expires: ${networkTokenCurrent.expires_at}, Simulated now: ${futureTime}\n`);
    }

    const discoveryTokenCurrent = client.getDiscoveryToken();
    if (discoveryTokenCurrent) {
      const expired = discoveryTokenCurrent.expires_at < futureTime;
      console.log(`Discovery token expired at simulated time: ${expired}`);
      console.log(`  Expires: ${discoveryTokenCurrent.expires_at}, Simulated now: ${futureTime}\n`);
    }

    if (domainId) {
      const domainAccessCurrent = client.getCachedDomainAccess(domainId);
      if (domainAccessCurrent) {
        const expired = domainAccessCurrent.expires_at < futureTime;
        console.log(`Domain access expired at simulated time: ${expired}`);
        console.log(`  Expires: ${domainAccessCurrent.expires_at}, Simulated now: ${futureTime}\n`);
      }
    }

    // Display Final State
    console.log('=== Final State ===');
    console.log(`Is authenticated: ${client.isAuthenticated()}`);

    const finalNetworkToken = client.getNetworkToken();
    if (finalNetworkToken) {
      console.log('\nNetwork Token:');
      console.log(`  Expires at: ${finalNetworkToken.expires_at}`);
      const remainingMs = finalNetworkToken.expires_at - now;
      console.log(`  Time remaining: ${remainingMs}ms (${Math.floor(remainingMs / 3600000)}h ${Math.floor((remainingMs % 3600000) / 60000)}m)`);
    }

    const finalDiscoveryToken = client.getDiscoveryToken();
    if (finalDiscoveryToken) {
      console.log('\nDiscovery Token:');
      console.log(`  Expires at: ${finalDiscoveryToken.expires_at}`);
      const remainingMs = finalDiscoveryToken.expires_at - now;
      console.log(`  Time remaining: ${remainingMs}ms (${Math.floor(remainingMs / 3600000)}h ${Math.floor((remainingMs % 3600000) / 60000)}m)`);
    }

    if (domainId) {
      const finalDomainAccess = client.getCachedDomainAccess(domainId);
      if (finalDomainAccess) {
        console.log('\nDomain Access:');
        console.log(`  Domain: ${finalDomainAccess.name} (${finalDomainAccess.id})`);
        console.log(`  Expires at: ${finalDomainAccess.expires_at}`);
        const remainingMs = finalDomainAccess.expires_at - now;
        console.log(`  Time remaining: ${remainingMs}ms (${Math.floor(remainingMs / 3600000)}h ${Math.floor((remainingMs % 3600000) / 60000)}m)`);
      }
    }

    console.log('\n=== Summary ===');
    console.log('• Network tokens have refresh_token and can be refreshed via POST /user/refresh');
    console.log('• Discovery tokens are obtained by re-authenticating with network token via POST /service/domains-access-token');
    console.log('• Domain access tokens are obtained by re-requesting with discovery token via POST /api/v1/domains/{id}/auth');
    console.log('• Only the network token can be truly "refreshed" - others must be re-requested\n');

    console.log('✓ Refresh test completed successfully!');

  } catch (error) {
    console.error('\n❌ Test failed:', error.message);
    if (error.retryable) {
      console.error('  This error is retryable');
    }
    process.exit(1);
  }
}

// Run the test
main().catch((error) => {
  console.error('💥 Unexpected error:', error);
  process.exit(1);
});

File: modules/authentication/pkg/python/examples/test_refresh.py
"""
Token Refresh Test

Tests the token refresh functionality of the authentication client.
Similar to examples/test_refresh.js but using the Python wrapper.
"""

import asyncio
import os
from dotenv import load_dotenv
from auki_authentication import Client, Config, current_time_ms

# Load environment variables
load_dotenv(dotenv_path='../../.env')

async def main():
    print('=== Token Refresh Test ===\n')

    # Configuration
    config = Config(
        api_url=os.getenv('API_URL', 'https://api.aukiverse.com'),
        refresh_url=os.getenv('REFRESH_URL', 'https://api.aukiverse.com/user/refresh'),
        dds_url=os.getenv('DDS_URL', 'https://dds.posemesh.org'),
        client_id=os.getenv('CLIENT_ID', 'python-test'),
        refresh_threshold_ms=int(os.getenv('REFRESH_THRESHOLD_MS', '300000'))
    )

    print('Configuration:')
    print(f'  API URL: {config.api_url}')
    print(f'  DDS URL: {config.dds_url}')
    print(f'  Client ID: {config.client_id}')
    print(f'  Refresh threshold: {config.refresh_threshold_ms}ms ({config.refresh_threshold_ms // 3600000}h)\n')

    # Credentials
    email = os.getenv('EMAIL')
    password = os.getenv('PASSWORD')

    if not email or not password:
        print('❌ EMAIL and PASSWORD must be set in .env file')
        return

    credentials = {
        'type': 'email',
        'email': email,
        'password': password,
    }

    print('Using Email/Password authentication')
    print(f'  Email: {email}\n')

    # Create the authentication client
    async with Client(config, credentials) as client:
        try:
            # Step 1: Initial Network Authentication
            print('=== Step 1: Initial Network Authentication ===')
            network_token = await client.authenticate()
            print('✓ Network token obtained!')
            print(f'  Token: {network_token.token[:50]}...')
            print(f'  Expires at: {network_token.expires_at} ({network_token.expires_at - current_time_ms()}ms from now)\n')

            # Step 2: Authenticate to Discovery
            print('=== Step 2: Discovery Authentication ===')
            discovery_token = await client.authenticate_discovery()
            print('✓ Discovery token obtained!')
            print(f'  Token: {discovery_token.token[:50]}...')
            print(f'  Expires at: {discovery_token.expires_at} ({discovery_token.expires_at - current_time_ms()}ms from now)\n')

            # Step 3: Get Domain Access
            domain_id = os.getenv('DOMAIN_ID')
            if not domain_id:
                print('⚠ DOMAIN_ID not set in .env, skipping domain access step\n')
            else:
                print('=== Step 3: Domain Access ===')
                print(f'Requesting access to domain: {domain_id}')
                domain_access = await client.get_domain_access(domain_id)
                print('✓ Domain access obtained!')
                print(f'  Domain: {domain_access.name} ({domain_access.id})')
                print(f'  Access token: {domain_access.access_token[:50]}...')
                print(f'  Expires at: {domain_access.expires_at} ({domain_access.expires_at - current_time_ms()}ms from now)\n')

            # Step 4: Test Token Refresh
            print('=== Step 4: Testing Token Refresh ===')
            print('Note: Only the network token has a refresh_token field.')
            print('Discovery and domain tokens must be re-requested, not refreshed.\n')

            print('Calling authenticate() again to test refresh logic...')

            # Get the current network token to compare
            network_token_before = client.get_network_token()
            print(f'Network token before: {network_token_before.token[:30]}...')

            # Try to authenticate again
            # Note: The Python wrapper's async API may handle this differently than JavaScript
            try:
                network_token_2 = await client.authenticate()
                print(f'Network token after: {network_token_2.token[:30]}...')
                print('✓ Token was refreshed or re-authenticated!\n')
            except Exception as error:
                # If no actions needed, wrapper may not make HTTP requests
                error_msg = str(error)
                if 'No response' in error_msg or 'No actions' in error_msg:
                    print('✓ No refresh needed - token is still valid and not near expiry\n')
                else:
                    print(f'✗ Unexpected error: {error_msg}')
                    raise

            # Step 5: Test Discovery Token Re-request
            print('=== Step 5: Testing Discovery Token Re-request ===')
            print('Calling authenticate_discovery() again...')

            try:
                discovery_token_2 = await client.authenticate_discovery()
                print('✓ Discovery token was re-requested!\n')
            except Exception as error:
                error_msg = str(error)
                if 'No response' in error_msg or 'No actions' in error_msg:
                    print('✓ No re-request needed - discovery token is still valid\n')
                else:
                    print(f'✗ Unexpected error: {error_msg}')
                    raise

            # Step 6: Simulate Token Expiration
            print('=== Step 6: Simulating Token Expiration ===')
            print('Simulating that 1 hour has passed (tokens should be expired)...\n')

            now = current_time_ms()
            future_time = now + 3660000  # 61 minutes

            print(f'Current time: {now}')
            print(f'Simulated time: {future_time} ({future_time - now}ms in the future)')
            print(f'Is authenticated: {client.is_authenticated()}\n')

            # Check which tokens are expired (relative to future time)
            network_token_current = client.get_network_token()
            if network_token_current:
                expired = network_token_current.expires_at < future_time
                print(f'Network token expired at simulated time: {expired}')
                print(f'  Expires: {network_token_current.expires_at}, Simulated now: {future_time}\n')

            discovery_token_current = client.get_discovery_token()
            if discovery_token_current:
                expired = discovery_token_current.expires_at < future_time
                print(f'Discovery token expired at simulated time: {expired}')
                print(f'  Expires: {discovery_token_current.expires_at}, Simulated now: {future_time}\n')

            if domain_id:
                domain_access_current = client.get_cached_domain_access(domain_id)
                if domain_access_current:
                    expired = domain_access_current.expires_at < future_time
                    print(f'Domain access expired at simulated time: {expired}')
                    print(f'  Expires: {domain_access_current.expires_at}, Simulated now: {future_time}\n')

            # Display Final State
            print('=== Final State ===')
            print(f'Is authenticated: {client.is_authenticated()}')

            final_network_token = client.get_network_token()
            if final_network_token:
                print('\nNetwork Token:')
                print(f'  Expires at: {final_network_token.expires_at}')
                remaining_ms = final_network_token.expires_at - now
                print(f'  Time remaining: {remaining_ms}ms ({remaining_ms // 3600000}h {(remaining_ms % 3600000) // 60000}m)')

            final_discovery_token = client.get_discovery_token()
            if final_discovery_token:
                print('\nDiscovery Token:')
                print(f'  Expires at: {final_discovery_token.expires_at}')
                remaining_ms = final_discovery_token.expires_at - now
                print(f'  Time remaining: {remaining_ms}ms ({remaining_ms // 3600000}h {(remaining_ms % 3600000) // 60000}m)')

            if domain_id:
                final_domain_access = client.get_cached_domain_access(domain_id)
                if final_domain_access:
                    print('\nDomain Access:')
                    print(f'  Domain: {final_domain_access.name} ({final_domain_access.id})')
                    print(f'  Expires at: {final_domain_access.expires_at}')
                    remaining_ms = final_domain_access.expires_at - now
                    print(f'  Time remaining: {remaining_ms}ms ({remaining_ms // 3600000}h {(remaining_ms % 3600000) // 60000}m)')

            print('\n=== Summary ===')
            print('• Network tokens have refresh_token and can be refreshed via POST /user/refresh')
            print('• Discovery tokens are obtained by re-authenticating with network token via POST /service/domains-access-token')
            print('• Domain access tokens are obtained by re-requesting with discovery token via POST /api/v1/domains/{id}/auth')
            print('• Only the network token can be truly "refreshed" - others must be re-requested\n')

            print('✓ Refresh test completed successfully!')

        except Exception as error:
            print(f'\n❌ Test failed: {error}')
            import traceback
            traceback.print_exc()
            return 1

    return 0

# Run the test
if __name__ == '__main__':
    exit_code = asyncio.run(main())
    exit(exit_code or 0)
